File: typescript/packages/bench/src/support/abstractPeriodicReport.ts
/** Base for reports that also run on a cadence. */

import { Cadence, type ScheduleContract } from "../contracts/scheduleContract";
import { AbstractReport } from "./abstractReport";

export abstract class AbstractPeriodicReport extends AbstractReport implements ScheduleContract {
    protected constructor(
        slug: string,
        title: string,
        readonly cadence: Cadence,
    ) {
        super(slug, title);
    }

    /** Next run instant, aligned to the cadence grid. */
    nextRunSeconds(now: number): number {
        const period = Cadence.seconds(this.cadence);

        return now - (now % period) + period;
    }
}

File: python/atelier_bench/support/abstract_periodic_report.py
"""Report that also runs on a cadence: adds ScheduleContract to the MRO."""

from __future__ import annotations

from ..contracts.schedule_contract import Cadence, ScheduleContract
from .abstract_report import AbstractReport


class AbstractPeriodicReport(AbstractReport, ScheduleContract):
    def __init__(self, slug: str, title: str, cadence: Cadence) -> None:
        super().__init__(slug, title)
        self.cadence = cadence
